editor/src/components/editor/actions/actions.tsx

```diff
@@ -3764,29 +3860,29 @@ export const UPDATE_FNS = {
       }
     }
   },
-  TRUE_UP_GROUPS: (editor: EditorModel): EditorModel => {
-    const targetsToTrueUp = editor.trueUpGroupsForElementAfterDomWalkerRuns.flatMap(
-      (trueUpTarget) => {
-        return trueUpTargetToTargets(editor.jsxMetadata, editor.elementPathTree, trueUpTarget)
-      },
+  TRUE_UP_ELEMENTS: (editor: EditorModel): EditorModel => {
+    const targets = mapDropNulls(
+      (target) => target,
+      editor.trueUpElementsAfterDomWalkerRuns.flatMap((trueUpTarget) => {
+        return trueUpTargetToPushIntendedBoundsTarget(
+          editor.jsxMetadata,
+          editor.elementPathTree,
+          trueUpTarget,
+        )
+      }),
     )
-    const canvasFrameAndTargets: Array<CanvasFrameAndTarget> = mapDropNulls((element) => {
-      const globalFrame = MetadataUtils.findElementByElementPath(
-        editor.jsxMetadata,
-        element,
-      )?.globalFrame
-      if (globalFrame == null || isInfinityRectangle(globalFrame)) {
-        return null
-      }
-      return {
-        frame: globalFrame,
-        target: element,
-      }
-    }, targetsToTrueUp)
-    const editorWithGroupsTruedUp = foldAndApplyCommandsSimple(editor, [
-      pushIntendedBoundsAndUpdateGroups(canvasFrameAndTargets, 'live-metadata'),
+
+    const groupTargets = targets.filter(isPushIntendedBoundsTargetGroup)
```

Comment:
is this extra step still needed? wouldn't it be better if `trueUpTargetToPushIntendedBoundsTarget` would instead return `Array<PushIntendedBoundsAndUpdateGroups | FixHuggingElements>`? is there some conceptual limtation I'm missing?

Reply:
I think returning commands here is a good refactor that improves readability, added that in [`3b83b7e` (#4359)](https://github.com/concrete-utopia/utopia/pull/4359/commits/3b83b7eb7f9b3a1db913972b97763c87ad5e9817).